### ISSUE TITLE
Style definition to allow relative paths in URLs

### ISSUE THREAD

**Issue by zstadler** [maplibre/maplibre-gl-js#182]
## Motivation

ArcGIS REST API has a [Vector Tile Style](https://developers.arcgis.com/rest/services-reference/enterprise/vector-tile-style.htm) service that returns styles for vector tiles **in Mapbox GL Style specification**.

While the [Mapbox Style](https://docs.mapbox.com/mapbox-gl-js/style-spec/root/) requires that the URLs for [`glyphs`](https://docs.mapbox.com/mapbox-gl-js/style-spec/root/#glyphs), [`sprite`](https://docs.mapbox.com/mapbox-gl-js/style-spec/root/#sprite), and `sources` must be absolute, the ArcGIS styles uses relative URLs. One advantage of this approach is that a style file can be used by different providers without modification.

It would be great if MapLibre is able to directly support style definitions with relative URLs.

An example from the above page:

```
{
   "version": 8,
   "sprite": "../sprites/sprite",
   "glyphs": "../fonts/{fontstack}/{range}.pbf",
   "sources": {
       "esri": {
           "type": "vector",
           "url": "../../"
       }
   },
   "layers": [
       {
           "id": "Graduated Color/GraduatedColor_poly/482025.000000 - 2135252.000000",
           "type": "fill",
           "source": "esri",
           "source-layer": "GraduatedColor_poly",
           "filter": [
               "==",
               "_symbol",
               0
           ],
           "layout": {},
           "paint": {
               "fill-color": "#FFFF80",
               "fill-outline-color": "#6E6E6E"
           }

...

       }
   ]
}
```

**Comment by xabbu42** [comment]
I think the idea is to interpret the paths relativ to the style as baseurl, instead of relativ to the current location maplibre-gl-js is running from. I am correct in my understanding @zstadler?

This actually seems to be the more useful way to handle relativ pathes and easy enough to implement. There could be installations out there which depend on the current behaviour of course, but as this styles are non-standard anyway I would not consider that change as breaking the api.

**Comment by zstadler** [comment]
Thanks @xabbu42 for taking this task in#645!

**Comment by klokan** [comment]
Discussed today on the SC meeting.

+1 for the idea - it seems to be a nice general implementation. Thanks for the effort!

Could we update the documentation in
https://github.com/maplibre/maplibre-gl-js-docs/tree/main/docs/pages/style-spec

Especially:
- https://maplibre.org/maplibre-gl-js-docs/style-spec/sources/
- https://maplibre.org/maplibre-gl-js-docs/style-spec/sprite/
- https://maplibre.org/maplibre-gl-js-docs/style-spec/glyphs/
to explicitly mention the absolute URL currently, and relative URL as an option proposed here?

**Comment by zstadler** [comment]
Keep alive

**Comment by animaux** [comment]
Will relative URLs be available in a specific version?

**Comment by HarelM** [comment]
I think this might be possible using `addProtocol` and/or `transformRequest`.
If so, a plugin that add this support would be a better solution to this issue, IMHO.
@animaux this is currently not available in a specific version.

**Comment by xabbu42** [comment]
As already discussed in #645, this can't be implemented with transformRequest.

**Comment by animaux** [comment]
> @animaux this is currently not available in a specific version.

Thanks, my question was more aiming at if an implementation of this is already planned for inclusion in a coming version.

**Comment by HarelM** [comment]
@Kai-W might have a solution for an issue similar to this one, maybe he could share what can be a workaround/solution to this issue as well.

**Comment by Kai-W** [comment]
I had urls relative to the pages origin (e.g. /fonts/{fontstack}/{range}.pbf)
transformRequest worked to transform the urls absolute paths:
`transformRequest: url => ({ url: location.origin + url})`

**Comment by animaux** [comment]
> As already discussed in #645, this can't be implemented with transformRequest.

@Kai-W thanks! So this is not true anymore?

This looks easy to implement. Where exactly does the `transformRequest` need to be included?

**Comment by xabbu42** [comment]
> > As already discussed in #645, this can't be implemented with transformRequest.
> 
> @Kai-W thanks! So this is not true anymore?

It is true for the general case. The described transformRequest uses the same base for all urls and therefore only works in specific scenarios. So you can work around the missing features for your situation where you now what urls to expect and what the correct base is. There is no way to write a transformRequest that can handle arbitrary styles with relative urls as they all may have different bases.

**Comment by KiwiKilian** [comment]
As this initial issue also talks about relative sprite URLs the problem of #3897 also applies, because the sprite url can't be rewritten, as it's validated before `transformRequest` is triggered.

**Comment by animaux** [comment]
@KiwiKilian Thanks, ok so this remains a problem.

**Comment by Kai-W** [comment]
yes i had problems with the sprite urls https://github.com/maplibre/maplibre-gl-js/issues/3897 the rest worked fine. A temporary fix https://github.com/maplibre/maplibre-gl-js/pull/3923 was merged a few days ago and a more permanet one https://github.com/maplibre/maplibre-gl-js/pull/3898 has a small breaking change and is linked to v5.

**Comment by KiwiKilian** [comment]
Great, thanks for the clarification @Kai-W!

**Comment by HarelM** [comment]
Ad described in #4962, This should be possible using `StyleTransformFunction` when setting a style, another suggestion would be to allow setting a `StyleTransformFunction` as part of the map options to facilitate for this. But this was not implemented yet, and I'm not sure it helps enough is this case for someone to implement it...

**Comment by animaux** [comment]
An absolute url to site root works well with `glyphs` by now:

``` js
"glyphs": "/map/fonts/{fontstack}/{range}.pbf",
```

If only this was possible with `sources/…/tiles`, like this:

``` js
    "sources": {
      "2023": {
        "type": "vector",
        "tiles": ["/map/2023/{z}/{x}/{y}.pbf"],
        …
      }
    },
```

This would make porting devsites a lot easier.
